File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/fetch/TaskSchedulers.java
package ai.platon.pulsar.crawl.fetch;

import ai.platon.pulsar.common.DateTimeUtil;
import ai.platon.pulsar.common.config.ImmutableConfig;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class TaskSchedulers implements AutoCloseable {

  public static final Logger LOG = LoggerFactory.getLogger(TaskSchedulers.class);

  private final String name;
  private Map<Integer, TaskScheduler> fetchSchedulers = Maps.newTreeMap();
  private Queue<Integer> fetchSchedulerIds = Lists.newLinkedList();

  public TaskSchedulers(ImmutableConfig conf) {
    this.name = this.getClass().getSimpleName() + "-" + DateTimeUtil.now("d.Hms");

    LOG.info("Initialize " + name);
  }

  public TaskSchedulers(List<TaskScheduler> taskSchedulers, ImmutableConfig conf) {
    this(conf);
    taskSchedulers.forEach(t -> put(t.getId(), t));
  }

  public synchronized String name() {
    return this.name;
  }

  public synchronized void put(int id, TaskScheduler taskScheduler) {
    fetchSchedulers.put(id, taskScheduler);
    fetchSchedulerIds.add(id);

    LOG.info("Add task scheduler #" + id);
    LOG.info("status : " + __toString());
  }

  public synchronized List<Integer> schedulerIds() {
    return new ArrayList<>(fetchSchedulerIds);
  }

  public synchronized TaskScheduler get(int id) {
    return fetchSchedulers.get(id);
  }

  public synchronized TaskScheduler getFirst() {
    return fetchSchedulers.values().iterator().next();
  }

  public synchronized TaskScheduler peek() {
    Integer id = fetchSchedulerIds.peek();
    if (id == null) {
      return null;
    }

    return fetchSchedulers.get(id);
  }

  public synchronized void remove(int id) {
    fetchSchedulerIds.remove(id);
    fetchSchedulers.remove(id);

    LOG.info("Remove FetchScheduler #" + id + " from pool #" + name());
    LOG.info("status : " + __toString());
  }

  public synchronized void remove(TaskScheduler taskScheduler) {
    if (taskScheduler == null) {
      return;
    }
    remove(taskScheduler.getId());
  }

  public synchronized void clear() {
    fetchSchedulerIds.clear();
    fetchSchedulers.clear();
  }

  /**
   * Random get @param count fetch items from an iterative selected job
   * */
  public synchronized List<FetchTask.Key> randomFetchItems(int count) {
    List<FetchTask.Key> keys = Lists.newArrayList();

    Integer id = fetchSchedulerIds.poll();
    if (id == null) {
      // log.debug("No running fetcher job");
      return Lists.newArrayList();
    }

    try {
      TaskScheduler taskScheduler = fetchSchedulers.get(id);
      if (taskScheduler == null) {
        LOG.error("Failed to find out the fetch scheduler with id #" + id);

        remove(id);
        return Lists.newArrayList();
      }

      for (FetchTask item : taskScheduler.schedule(count)) {
        keys.add(item.getKey());
      }
    }
    catch (Throwable e) {
      LOG.error(e.toString());
    }

    fetchSchedulerIds.add(id); // put back to the queue

    return keys;
  }

  @Override
  public synchronized String toString() {
    return __toString();
  }

  private String __toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("Job IDs : ")
        .append(StringUtils.join(fetchSchedulerIds, ", "))
        .append("\tQueue Size : ")
        .append(fetchSchedulers.size());

    return sb.toString();
  }

  @Override
  public void close() throws Exception {
    LOG.info("[Destruction] Closing TaskSchedulers");

    fetchSchedulerIds.clear();
    fetchSchedulers.clear();
  }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/fetch/TaskSchedulers.kt
package ai.platon.pulsar.crawl.fetch

import ai.platon.pulsar.common.DateTimes
import ai.platon.pulsar.common.config.ImmutableConfig
import com.google.common.collect.Lists
import com.google.common.collect.Maps
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import java.util.*

class TaskSchedulers(conf: ImmutableConfig) : AutoCloseable {
    val name: String = this.javaClass.simpleName + "-" + DateTimes.now("d.Hms")
    private val fetchSchedulers = Maps.newTreeMap<Int, TaskScheduler>()
    private val fetchSchedulerIds = Lists.newLinkedList<Int>()

    val first: TaskScheduler
        @Synchronized get() = fetchSchedulers.values.iterator().next()

    constructor(taskSchedulers: List<TaskScheduler>, conf: ImmutableConfig) : this(conf) {
        taskSchedulers.forEach { t -> put(t.id, t) }
    }

    @Synchronized
    fun put(id: Int, taskScheduler: TaskScheduler) {
        fetchSchedulers[id] = taskScheduler
        fetchSchedulerIds.add(id)

        LOG.info("Add task scheduler #$id")
        LOG.info("status : " + __toString())
    }

    @Synchronized
    fun schedulerIds(): List<Int> {
        return ArrayList(fetchSchedulerIds)
    }

    @Synchronized
    operator fun get(id: Int): TaskScheduler? {
        return fetchSchedulers[id]
    }

    @Synchronized
    fun peek(): TaskScheduler? {
        val id = fetchSchedulerIds.peek() ?: return null
        return fetchSchedulers[id]
    }

    @Synchronized
    fun remove(id: Int) {
        fetchSchedulerIds.remove(id)
        fetchSchedulers.remove(id)

        LOG.info("Remove FetchScheduler #$id from pool #$name")
        LOG.info("status : " + __toString())
    }

    @Synchronized
    fun remove(taskScheduler: TaskScheduler?) {
        if (taskScheduler == null) {
            return
        }
        remove(taskScheduler.id)
    }

    @Synchronized
    fun clear() {
        fetchSchedulerIds.clear()
        fetchSchedulers.clear()
    }

    /**
     * Random get @param count fetch items from an iterative selected job
     */
    @Synchronized
    fun randomFetchItems(count: Int): List<JobFetchTask.Key> {
        val keys = mutableListOf<JobFetchTask.Key>()
        val id = fetchSchedulerIds.poll()?: return listOf()

        try {
            val taskScheduler = fetchSchedulers[id]
            if (taskScheduler == null) {
                LOG.error("Failed to find out the fetch scheduler with id #$id")
                remove(id)
                return listOf()
            }

            taskScheduler.schedule(count).mapTo(keys) { it.key }
        } catch (e: Throwable) {
            LOG.error("Unexpected exception", e)
        }

        // put back to the queue
        fetchSchedulerIds.add(id)

        return keys
    }

    @Synchronized
    override fun toString(): String {
        return __toString()
    }

    private fun __toString(): String {
        val sb = StringBuilder()
        sb.append("Job IDs : ")
                .append(StringUtils.join(fetchSchedulerIds, ", "))
                .append("\tQueue Size : ")
                .append(fetchSchedulers.size)

        return sb.toString()
    }

    @Throws(Exception::class)
    override fun close() {
        LOG.info("[Destruction] Closing TaskSchedulers")

        fetchSchedulerIds.clear()
        fetchSchedulers.clear()
    }

    companion object {
        val LOG = LoggerFactory.getLogger(TaskSchedulers::class.java)
    }
}
